### ISSUE TITLE
Proper handling of random state for models that use randomness (i.e. dropout) is needed for some optimizers

### ISSUE THREAD

**Issue by crowsonkb** [pytorch/functorch#408]
There are a number of optimization methods that require you to be able to evaluate a pure function over and over at different inputs (parameters to optimize over) but there is one common pitfall when using them, namely that if the model or loss function uses randomness then you have to be able to call it with the *same* randomness over and over, for instance, the same features dropped out. Using different randomness will break most common line search algorithms, for instance, pretty easily. Currently the way I would have to do this is manually fork the RNG, even with functorch. A good way to manage random state/wrap forking the RNG into a pure function that takes a seed as one of it arguments/some such thing would be much appreciated!

Thank you,
Katherine Crowson

**Comment by samdow** [comment]
Hi @crowsonkb! Thanks for the issue! In an attempt to delay some heavy duty refactors to the underlying PyTorch, we've been working on adding a flag version of this (i.e. `vmap(...., different_randomess=False)` with the name being up for discussion). I'd been trying to wait until everything is ready to go to put up the diffs but now I'll put up what I have so far since the remaining operations are just distributions.

One question for you: Do you have cases where within the same vmap call you sometimes want the same randomness and sometimes want the different randomness behavior? As a concrete example, would there be a case where during a single run of vmap you would want some dropouts to use the same randomness and others not to? If there is the flag version won't be able to solve this, so this will help us reprioritize!

**Comment by crowsonkb** [comment]
I wouldn't actually be using vmap for this specific thing, I think. It would be a normal evaluation of a model with a normal batch dimension, I just have to use the same randomness on subsequent calls (not done at the same time, for instance an iterative line search with a termination criterion). I can't think of anything offhand where I would want to repeat some but not all randomness in a vmap but I'm sure I would run into it at some point...

**Comment by samdow** [comment]
Thanks for that context! To unblock you if you aren't using vmap, does it work to [reset the manual seed](https://discuss.pytorch.org/t/does-random-seed-in-pytorch-has-the-effect-on-the-function-of-drop-out/24741)? If not, I would be curious why it's not working

**Comment by samdow** [comment]
Hi @crowsonkb! Just waned to let you know that "different" and "same" randomness are supported in vmap in our 0.1 release that came out alongside PyTorch 1.11. [More documentation here](https://pytorch.org/functorch/nightly/ux_limitations.html#randomness), sorry it's on nightly but is definitely in 1.11

If you have a chance to play around with it, any feedback is really appreciated as we're solidifying our APIs and trying to understand what use cases this covers and does not cover
